File: julia/src/info.jl
# What a file says about itself, printed.
#
# Section 29 asks that opening a file read no array, and this is what
# that buys: everything below comes from attributes and dataspaces.
# What is printed is the list of `docs/api-conventions.md` section 5,
# in the same order in every language: for a key its name, role,
# units, bounds, category, trajectory group and parent; for a support
# its kind, its counts and its id; for a slot its shape with the name
# of each axis, its units and its source, and for a callable slot the
# callable and the output it takes.

"""
    info(path; io = stdout)
    info(ds; io = stdout)

Print what a file declares, without reading an array.  A semantic
fault never stops this: a file with a missing unit or a broken split
is exactly the file a reader most needs to look at.

    Mestra.info("case.mes")
"""
function info(path::AbstractString; io::IO = stdout, kwargs...)
    ds = read(String(path); lazy = true, kwargs...)
    return info(ds; io = io, name = String(path))
end

function info(ds::Dataset; io::IO = stdout, name = nothing)
    head = something(name, ds.path, "dataset")
    println(io, head, ": ", ds.format, ", ", ds.nrows, " row(s), ",
            isempty(ds.supports) ? "no support" :
            ds.aligned ? "aligned" : "unaligned",
            ", written by ", ds.writer, " at ", ds.created)
    ds.generalisation_group === nothing ||
        println(io, "  unit of generalisation: ",
                ds.generalisation_group)

    if !isempty(ds.keys)
        println(io, "keys")
        names = key_order(ds)
        w = maximum(length.(names))
        for n in names
            println(io, "  ", rpad(n, w), "  ", key_line(ds, ds.keys[n]))
        end
    end

    if !isempty(ds.categories)
        println(io, "categories")
        for t in sort(collect(Base.keys(ds.categories)), by = codeunits)
            e = ds.categories[t].entries
            println(io, "  ", t, "  ", length(e), ": ",
                    join(e[1:min(6, end)], ", "),
                    length(e) > 6 ? ", ..." : "")
        end
    end

    if !isempty(ds.supports)
        println(io, "supports")
        w = maximum(length(s.name) for s in ds.supports)
        for s in support_order(ds)
            println(io, "  ", rpad(s.name, w), "  ", rpad(s.kind, 5), "  ",
                    s.n_nodes, " node(s), ", s.n_cells, " cell(s)  ",
                    s.support_id)
        end
    end

    slots = all_slots(ds)
    if !isempty(slots)
        println(io, "slots")
        w = maximum(length(s.path) for s in slots)
        for s in slots
            println(io, "  ", rpad(s.path, w), "  ", slot_line(s))
        end
    end

    if !isempty(ds.callables)
        println(io, "callables")
        for id in sort(collect(Base.keys(ds.callables)), by = codeunits)
            c = ds.callables[id]
            println(io, "  ", id, "  type ", something(c.type, "-"),
                    c.repr === nothing ? "" : "  " * c.repr)
        end
    end

    if !isempty(ds.findings)
        println(io, "what the reader would not follow or could not read")
        for f in ds.findings
            println(io, "  ", f)
        end
    end
    return ds
end

"""One key: the fields of section 5, and the ones it does not carry
are left out rather than printed as blanks."""
function key_line(ds::Dataset, k::KeyColumn)
    parts = String[rpad(k.role === nothing ? "-" : String(k.role), 11)]
    k.units === nothing || push!(parts, "units " * k.units)
    (k.lower === nothing || k.upper === nothing) ||
        push!(parts, "bounds [" * string(k.lower) * ", " *
                     string(k.upper) * "]")
    k.category === nothing || push!(parts, "category " * k.category)
    k.trajectory_group === nothing ||
        push!(parts, "trajectory_group " * k.trajectory_group)
    k.parent === nothing || push!(parts, "parent " * k.parent)
    return join(parts, "  ")
end

"""One slot: the shape with the name of every axis, then the units and
where the numbers come from."""
function slot_line(s::Slot)
    parts = String[]
    if !isempty(s.ldims)
        push!(parts, "(" * join(string.(s.ldims), ", ") * ") " *
                     join(s.dshape, "x"))
    elseif s.components !== nothing
        push!(parts, "(row, " *
                     (s.location === :cell ? "cell" : "node") *
                     ", component) " * string(s.components) *
                     " component(s)")
    end
    s.role === nothing || push!(parts, String(s.role))
    push!(parts, "units " * something(s.units, "-"))
    s.statistic === nothing || push!(parts, "statistic " * s.statistic)
    push!(parts, is_callable_slot(s) ?
          "callable " * callable_id(s) * " -> " * something(s.output, "-") :
          "data")
    s.recomputed === true && push!(parts, "recomputed")
    return join(parts, "  ")
end

File: matlab/+mestra/info.m
function out = info(path, varargin)
%MESTRA.INFO  Print what is in a file, without reading any array.
%
%   MESTRA.INFO(PATH) prints the fields every implementation prints
%   (docs/api-conventions.md, section 5):
%
%     * for every key, its name, role, units, bounds, category,
%       trajectory group and parent;
%     * for every support, its kind, its node and cell counts and its
%       id;
%     * for every slot, its shape with the axes named, its units, its
%       source, and for a callable slot the callable id and the
%       output it fills.
%
%   It goes through MESTRA.OPEN, so it reads attributes and
%   dataspaces and no array at all (specification section 29).  That
%   is what makes it usable on a file too large to read, and on the
%   files a user most needs to inspect.
%
%       mestra.info('mesh_two_rows.mes')
%       mesh_two_rows.mes
%         format mestra/0, writer mestra python 0
%         created 2026-09-19T00:00:00Z
%         2 row(s), aligned, unit of generalisation member
%
%         keys
%           mach     condition  units 1     bounds [0.4, 0.8]
%           member   group      category member
%         ...
%
%   A shape is printed in the file's own axis order with every axis
%   named, `row=2 node=6 component=1`, because that is the order the
%   file states.  What MESTRA.READ hands back in MATLAB is the
%   reverse of it; permute by name with MESTRA.PERMUTE.
%
%   S = MESTRA.INFO(PATH, 'String', true) returns the text instead of
%   printing it.
%
%   MESTRA.INFO does not validate.  MESTRA.VALIDATE and MESTRA.REPORT
%   do that, and a file with a semantic fault still prints here,
%   which is the point.
%
%   See also mestra.open, mestra.validate, mestra.report, mestra.read.

    p = inputParser();
    p.addParameter('String', false);
    p.addParameter('File', 1);
    p.parse(varargin{:});

    d = mestra.open(path, 'Strict', false);
    L = {};

    [~, base, ext] = fileparts(path);
    L{end + 1} = [base ext];
    L{end + 1} = sprintf('  format %s, writer %s', d.format, d.writer);
    L{end + 1} = sprintf('  created %s', d.created);
    line = sprintf('  %d row(s)', d.nRows);
    if ~isempty(d.supports)
        if d.aligned
            line = [line ', aligned'];
        else
            line = [line ', unaligned'];
        end
    end
    if ~isempty(d.generalisationGroup)
        line = [line ', unit of generalisation ' d.generalisationGroup];
    end
    L{end + 1} = line;

    if ~isempty(d.keys)
        L{end + 1} = '';
        L{end + 1} = '  keys';
        width = max(cellfun(@numel, {d.keys.name}));
        for i = 1:numel(d.keys)
            k = d.keys(i);
            bits = {};
            if ~isempty(k.units)
                bits{end + 1} = ['units ' k.units]; %#ok<AGROW>
            end
            if ~isempty(k.lower) || ~isempty(k.upper)
                bits{end + 1} = sprintf('bounds [%s, %s]', ...
                    num(k.lower), num(k.upper)); %#ok<AGROW>
            end
            if ~isempty(k.category)
                bits{end + 1} = ['category ' k.category]; %#ok<AGROW>
            end
            if ~isempty(k.trajectoryGroup)
                bits{end + 1} = ['trajectory group ' ...
                                 k.trajectoryGroup]; %#ok<AGROW>
            end
            if ~isempty(k.parent)
                bits{end + 1} = ['parent ' k.parent]; %#ok<AGROW>
            end
            L{end + 1} = sprintf('    %-*s  %-11s %s', width, k.name, ...
                                 k.role, strjoin(bits, '  ')); %#ok<AGROW>
        end
    end

    if ~isempty(d.categories)
        L{end + 1} = '';
        L{end + 1} = '  category tables';
        for i = 1:numel(d.categories)
            c = d.categories(i);
            L{end + 1} = sprintf('    %s  %d entr%s: %s', c.name, ...
                numel(c.entries), plural(numel(c.entries)), ...
                strjoin(c.entries, ', ')); %#ok<AGROW>
        end
    end

    if ~isempty(d.scalars)
        L{end + 1} = '';
        L{end + 1} = '  scalars';
        for i = 1:numel(d.scalars)
            L{end + 1} = ['    ' slotLine(d.scalars(i), {'row'}, ...
                                          d.nRows)]; %#ok<AGROW>
        end
    end

    if ~isempty(d.supports)
        L{end + 1} = '';
        L{end + 1} = '  supports';
        for i = 1:numel(d.supports)
            s = d.supports(i);
            L{end + 1} = sprintf('    %s  kind %s, %d node(s), %d cell(s)', ...
                                 s.name, s.kind, s.nNodes, s.nCells); %#ok<AGROW>
            L{end + 1} = sprintf('      id %s', s.supportId); %#ok<AGROW>
            if ~isempty(s.coordinates)
                L{end + 1} = ['      ' slotLine(s.coordinates)]; %#ok<AGROW>
            end
            for j = 1:numel(s.nodeArrays)
                L{end + 1} = ['      node ' ...
                              slotLine(s.nodeArrays(j))]; %#ok<AGROW>
            end
            for j = 1:numel(s.cellArrays)
                L{end + 1} = ['      cell ' ...
                              slotLine(s.cellArrays(j))]; %#ok<AGROW>
            end
        end
    end

    if ~isempty(d.callables)
        L{end + 1} = '';
        L{end + 1} = '  callables';
        for i = 1:numel(d.callables)
            c = d.callables(i);
            line = sprintf('    %s  type %s', c.id, c.type);
            if ~isempty(c.repr)
                line = [line '  ' c.repr]; %#ok<AGROW>
            end
            L{end + 1} = line; %#ok<AGROW>
        end
    end

    if ~isempty(d.skipped)
        L{end + 1} = '';
        L{end + 1} = '  passed over';
        for i = 1:numel(d.skipped)
            L{end + 1} = ['    ' d.skipped{i}]; %#ok<AGROW>
        end
    end

    text = [strjoin(L, newline) newline];
    if p.Results.String
        out = text;
        return
    end
    fprintf(p.Results.File, '%s', text);
    if nargout > 0
        out = d;
    end
end

function s = slotLine(slot, dims, rowCount)
%slotLine  One slot: its name, its shape with the axes named, its
%   units and its source.
    if nargin < 2, dims = slot.dims; end
    bits = {};
    shape = '';
    if isfield(slot, 'shape') && ~isempty(slot.shape)
        shape = namedShape(fliplr(dims), slot.shape);
    elseif nargin >= 3
        shape = sprintf('row=%d', rowCount);
    end
    if ~isempty(shape)
        bits{end + 1} = ['(' shape ')'];
    end
    if isfield(slot, 'role') && ~isempty(slot.role)
        bits{end + 1} = slot.role;
    end
    if ~isempty(slot.units)
        bits{end + 1} = ['units ' slot.units];
    end
    if isfield(slot, 'varies') && ~isempty(slot.varies)
        bits{end + 1} = ['varies ' slot.varies];
    end
    if isfield(slot, 'components') && ~isempty(slot.components)
        bits{end + 1} = sprintf('components %d', slot.components);
    end
    if isfield(slot, 'category') && ~isempty(slot.category)
        bits{end + 1} = ['category ' slot.category];
    end
    if isfield(slot, 'statistic') && ~isempty(slot.statistic)
        bits{end + 1} = ['statistic ' slot.statistic];
    end
    src = slot.source;
    if numel(src) > 9 && strncmp(src, 'callable:', 9)
        bits{end + 1} = ['source callable ' src(10:end)];
        if ~isempty(slot.output)
            bits{end + 1} = ['output ' slot.output];
        end
    else
        bits{end + 1} = ['source ' src];
    end
    s = sprintf('%s  %s', slot.name, strjoin(bits, ', '));
end

function s = namedShape(fileDims, shape)
%namedShape  `row=2 node=6 component=1`, in the file's axis order.
    n = min(numel(fileDims), numel(shape));
    bits = cell(1, n);
    for i = 1:n
        bits{i} = sprintf('%s=%d', fileDims{i}, shape(i));
    end
    s = strjoin(bits, ' ');
end

function s = num(v)
    if isempty(v)
        s = '-';
    else
        s = strtrim(sprintf('%g', double(v)));
    end
end

function s = plural(n)
    if n == 1, s = 'y'; else, s = 'ies'; end
end
